src/AgreementManager.ts

```diff
@@ -308,49 +278,54 @@ class AgreementManager {
         encodedLatestCorrectState: string;
         virtualVotingBlocks: ConfirmedBlockStruct[];
     } {
-        let fork = this.forks[Number(forkCnt)];
-        let encodedLatestFinalizedState: string | undefined;
-        let encodedLatestCorrectState: string | undefined;
-        let virtualVotingBlocks: ConfirmedBlockStruct[] = [];
-        let set = new Set(fork.addressesInThreshold);
-        for (let i = fork.agreements.length - 1; i >= 0; i--) {
-            let agreement = fork.agreements[i];
-            let signerAddresses: AddressLike[] = [];
-            for (let signature of agreement.blockSignatures) {
-                let adr = EvmUtils.retrieveSignerAddressBlock(
-                    agreement.block,
-                    signature
-                );
-                if (!encodedLatestCorrectState && adr == signerAddress) {
-                    encodedLatestCorrectState = agreement.encodedState;
+        const fork = this.forks[Number(forkCnt)];
+
+        const result = fork.agreements.reduceRight(
+            (acc, agreement, index) => {
+                // Only update encodedLatestCorrectState if not found yet and this participant signed
```

Comment:
This runs for every element in the array and does not take into account virtual voting on the signatures. The original implementation even though it starts from the last agreement and it seems it runs to the first, it returns (breaks) the moment a threshold is reached through virtual voting (look how the set is modified) - so if the leader election algorithm is round robin this is at most N runs, where N is the number of participants.

Reply:
OK ill double checked and add test to assert this behivior

Reply:
fixed and tested https://github.com/peer3to/state-channels-plus/pull/21/commits/ca129eff3e78cbc7662b6d25de36a41d4dbbf115

Reply:
Not sure what you mean with that. That commit modifies 'didEveryoneSignBlock' - that function doesn't have anything to do with virtual voting and the check you added there (set.size == 0) doesn't change anything (unless signatures from non participants are stored which isn't the case) 

My comment was on this function 'getFinalizedAndLatestWithVotes' and specifically 'reduceRight' - it will always run on all elements of the array and the logic doesn't take into account virtual voting - take a look at the removed code of that function